crates/scope/src/builder.rs

```diff
@@ -2655,9 +2687,80 @@ impl ScopeDataMapBuilder {
             enclosing,
             body_scope_builder,
         );
-
         self.possibly_annex_b_functions.clear();
 
+        let has_extra_body_var = match &scope_data_set.extra_body_var {
+            ScopeData::Var(_) => true,
+            _ => false,
+        };
+
+        let fun_stencil = self.function_stencil_builder.current_mut();
+
+        if let ScopeData::Function(fun) = &scope_data_set.function {
+            if fun.base.needs_environment_object() {
+                fun_stencil.set_needs_environment_object();
+            }
+        } else {
+            panic!("Unexpected scope data for function");
+        }
+
+        if has_extra_body_var {
+            fun_stencil.set_has_extra_body_var_scope();
+        }
+
+        if has_mapped_arguments {
+            fun_stencil.set_has_mapped_args_obj();
+        }
+
+        if !fun_stencil.is_arrow() {
+            if has_used_this {
+                fun_stencil.set_has_this_binding();
+            }
+
+            // Perform the same algorithm as
+            // ParseContext::declareFunctionArgumentsObject in
+            // m-c/js/src/frontend/ParseContext.cpp.
+            //
+            // NOTE: This is implementation-specfic optimization, and has
+            //       no corresponding steps in the spec.
+
+            let mut uses_arguments = false;
+            let mut try_declare_arguments = has_used_arguments;
+            if has_extra_body_var {
+                if var_names_has_arguments {
+                    try_declare_arguments = true;
+                }
+            } else {
+                if var_names_has_arguments && !parameter_has_arguments {
+                    uses_arguments = true;
+                }
+            }
+
+            if try_declare_arguments {
+                if has_extra_body_var {
+                    if !parameter_has_arguments {
+                        fun_stencil.set_should_declare_arguments();
+                        uses_arguments = true;
+                    }
+                } else {
+                    let has_defined_arguments =
+                        parameter_has_arguments || body_has_defined_arguments;
+
+                    if !has_defined_arguments {
+                        fun_stencil.set_should_declare_arguments();
+                        uses_arguments = true;
+                    }
+                }
```

Comment:
nit: This code could be factored to avoid repeating the 2 lines.
```rust
let has_defined_arguments = parameter_has_arguments || body_has_defined_arguments;
let declare_arguments = if has_extra_body_var { !parameter_has_arguments } else { has_defined_arguments };
if declare_arguments {
  fun_stencil.set_should_declare_arguments();
  uses_arguments = true;
}
```

Also, please add back the comments from the original code, that argument is shadowed by the parameters.